### PR TITLE
Setup api tests

### PR DESCRIPTION
**Pull request by mscibilia**
Created tests for REST endpoints using RestAssured.
#18

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -11,6 +11,11 @@ ext {
     apacheCommonsCliVersion = '1.4'
     jsonVersion = '20180130'
     slf4jVersion = '1.7.25'
+    restAssuredVersion = '4.1.1'
+    restAssuredKotlinExtVersion = '4.1.2'
+    junitJupiterVersion = '5.5.2'
+    mockkVersion = '1.9.3'
+    jacksonModuleKotlinVersion = '2.10.0.pr3'
```
Comment (Edvinas01) [inline build.gradle]: The latest version seems to be `2.10.0` (without the `pr` suffix):
https://mvnrepository.com/artifact/com.fasterxml.jackson.module/jackson-module-kotlin/2.10.0

File: build.gradle [hunk 2]
```diff
@@ -35,6 +40,13 @@ dependencies {
     // Logging.
     compile "org.slf4j:slf4j-log4j12:$slf4jVersion"
     compile "org.slf4j:slf4j-api:$slf4jVersion"
+
+    // Test
+    testImplementation "io.rest-assured:rest-assured:$restAssuredVersion"
+    testImplementation "io.rest-assured:kotlin-extensions:$restAssuredKotlinExtVersion"
+    testImplementation "org.junit.jupiter:junit-jupiter:$junitJupiterVersion"
```
Comment (Edvinas01) [inline build.gradle]: JUnit 5 :+1:

File: src/test/kotlin/com/edd/memegrid/web/RouterTest.kt [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
+    @BeforeEach
+    fun beforeEach() {
+        clearMocks(mockImageValidator, mockMemeManager, mockTemplateReader)
+    }
+    @Test
+    fun `should return 200 given get request with HTML content type`() {
+        Given {
+            port(PORT)
```
Comment (Edvinas01) [inline src/test/kotlin/com/edd/memegrid/web/RouterTest.kt]: You can assign a default port in `@BeforeAll` - this way you can avoid writing `port(PORT)` in each test. For example:
```kotlin
@BeforeAll
fun startServer() {
    ...

    RestAssured.port = PORT
}
```

File: src/test/kotlin/com/edd/memegrid/web/RouterTest.kt [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+            accept(ContentType.JSON)
+            body(meme)
+        } When {
+            post("/api/memes")
+        } Then {
+            statusCode(STATUS_CODE_SERVER_ERROR)
+        }
+    }
+
+    private fun JsonNode.toMemeList() = jacksonObjectMapper().readValue<List<Meme>>(this.toPrettyString())
```
Comment (Edvinas01) [inline src/test/kotlin/com/edd/memegrid/web/RouterTest.kt]: I'd suggest to place `jacksonObjectMapper()` inside the `companion object`, assign it to a `val` and reference it here. As `jacksonObjectMapper` creates a new instance of `ObjectMapper` each time it is called:
```kotlin
fun jacksonObjectMapper(): ObjectMapper = ObjectMapper().registerKotlinModule()
```

File: build.gradle (hunk 1 above)
Comment (mscibilia) [reply build.gradle]: Changed to 2.10.0

File: src/test/kotlin/com/edd/memegrid/web/RouterTest.kt (hunk 4 above)
Comment (mscibilia) [reply src/test/kotlin/com/edd/memegrid/web/RouterTest.kt]: Moved to companion object

File: src/test/kotlin/com/edd/memegrid/web/RouterTest.kt (hunk 3 above)
Comment (mscibilia) [reply src/test/kotlin/com/edd/memegrid/web/RouterTest.kt]: port is now set in beforeAll

**Comment by mscibilia** [comment]
> Thank you for the PR! Most of it looks good except for few minor details. Also #38 migrated the Gradle file to Kotlin DSL which has few conflicts with your PR, so that will have to be updated.

Thanks! I have completed the changes and converted mine to Kotlin DSL also :)

**Pull request merged**